lib/imagecanvas.cpp

```diff
@@ -2108,25 +2214,26 @@ void ImageCanvas::applyCurrentTool()
             if (filledImage.isNull())
                 return;
 
-            mProject->beginMacro(QLatin1String("PixelTexturedFillTool"));
-            mProject->addChange(new ApplyPixelFillCommand(this, mProject->currentLayerIndex(),
-                *currentProjectImage(), filledImage));
-            // mProject->endMacro();
+            command = new ApplyPixelFillCommand(this, mProject->currentLayerIndex(), *currentProjectImage(), filledImage);
+            command->setText(QLatin1String("PixelTexturedFillTool"));
         } else {
             const QImage filledImage = greedyFillPixels();
             if (filledImage.isNull())
                 return;
 
-            mProject->beginMacro(QLatin1String("GreedyPixelTexturedFillTool"));
-            mProject->addChange(new ApplyGreedyPixelFillCommand(this, mProject->currentLayerIndex(),
-                *currentProjectImage(), filledImage));
-            // mProject->endMacro();
+            command = new ApplyGreedyPixelFillCommand(this, mProject->currentLayerIndex(), *currentProjectImage(), filledImage);
+            command->setText(QLatin1String("GreedyPixelTexturedFillTool"));
         }
         break;
     }
     default:
         break;
     }
+
+    if (command) {
+//        mProject->addChange(command);
+        stack->push(command);
```

Comment:
Although it does make the code uglier, it would be nice if mProject->addChange() could be called in the case of the stack being project->undoStack(), as it allows the logging category to do its thing, which can be useful when debugging. i.e:

if (alternateStack)
    alternateStack->push(command);
else
    mProject->addChange(command);